Asynchronous TS plugin example

johnsoncodehk opened the pull request:
See https://github.com/vuejs/language-tools/discussions/3789

Different from vue langauge tools, mdx-analyzer needs to load the plugin asynchronously, so `decorateLanguageService` and `decorateLanguageServiceHost` are called lazily.

johnsoncodehk committed:
Asynchronous TS plugin example

remcohaszing commented on packages/vscode-mdx/src/typescript-plugin.js:
```diff
@@ -0,0 +1,107 @@
+/**
+ * @type {PluginModuleFactory} modules
+ */
+const init = (modules) => {
```
```suggestion
function init(modules) {
```

remcohaszing commented on packages/vscode-mdx/src/typescript-plugin.js:
```diff
@@ -0,0 +1,107 @@
+  const {typescript: ts} = modules
+  return {
+    create(info) {
+      const enptySnapshot = ts.ScriptSnapshot.fromString('')
```
```suggestion
      const emptySnapshot = ts.ScriptSnapshot.fromString('')
```

remcohaszing commented on packages/vscode-mdx/src/typescript-plugin.js:
```diff
@@ -0,0 +1,107 @@
+      }
+      info.languageServiceHost.getScriptVersion = (fileName) => {
+        if (!initialized && fileName.endsWith('.mdx')) {
+          return 'initializing...'
```
I think this can be any string, as long as it’s not numeric, right?

```suggestion
          return 'initializing…'
```

johnsoncodehk replied:
Yes.

remcohaszing commented on packages/vscode-mdx/src/typescript-plugin.js:
```diff
@@ -0,0 +1,107 @@
+      if (getProjectVersion) {
+        info.languageServiceHost.getProjectVersion = () => {
+          if (!initialized) {
+            return getProjectVersion() + ',initializing...'
```
I think this is rendered in the VS Code status bar, right?

Let’s use a real ellipsis and proper spacing.

```suggestion
            return getProjectVersion() + ', initializing…'
```

johnsoncodehk replied:
This will not be shown to the user. The purpose of patching getProjectVersion is just so that after the initialization is completed, the language service knows that the cache needs to be refreshed.

remcohaszing commented on packages/vscode-mdx/src/typescript-plugin.js:
```diff
@@ -0,0 +1,107 @@
+  }
+}
+
+exports = init
```
Reassigning `exports` doesn’t do anything. That means this module is entirely unused.

I think you’re confusing the following TypeScript syntax:

```ts
export = init
```

which compiles to the following JavaScript:

```suggestion
module.exports = init
```

remcohaszing commented on packages/vscode-mdx/src/typescript-plugin.js:
```diff
@@ -0,0 +1,107 @@
+/**
+ * @typedef {import('typescript/lib/tsserverlibrary.js').server.Project} Project
+ * @typedef {import('typescript/lib/tsserverlibrary.js').server.PluginModuleFactory} PluginModuleFactory
+ * @typedef {import('typescript/lib/tsserverlibrary.js').LanguageService} LanguageService
```
This is equivalent nowadays.

```suggestion
 * @typedef {import('typescript').server.Project} Project
 * @typedef {import('typescript').server.PluginModuleFactory} PluginModuleFactory
 * @typedef {import('typescript').LanguageService} LanguageService
```

remcohaszing commented on packages/vscode-mdx/package.json:
```diff
@@ -186,6 +183,16 @@
+    ],
+    "typescriptServerPlugins": [
+      {
+        "name": "typescript-plugin-bundled",
```
Does this have to come from `node_modules`? Or can we also point to a relative path?

johnsoncodehk replied:
It must come from `node_modules` and can't be a relative path.

- https://github.com/microsoft/vscode/blob/46a233646602c378c2168fa2a104dff8e8551558/extensions/typescript-language-features/src/tsServer/spawner.ts#L252
- https://github.com/microsoft/TypeScript/blob/369e161e4c37c5f166c64973e7f1217c2298d7dd/src/server/editorServices.ts#L4491

remcohaszing commented on packages/vscode-mdx/src/typescript-plugin.js:
```diff
@@ -0,0 +1,107 @@
+
+      info.languageServiceHost.getScriptSnapshot = (fileName) => {
+        if (!initialized && fileName.endsWith('.mdx')) {
+          return enptySnapshot
```
```suggestion
          return emptySnapshot
```

remcohaszing commented on packages/vscode-mdx/src/typescript-plugin.js:
```diff
@@ -0,0 +1,107 @@
+} from '@volar/typescript'
+import {createFileProvider, resolveCommonLanguageId} from '@volar/language-core'
+import {getLanguageModule} from '@mdx-js/language-service'
+import {loadPlugins} from '../../language-server/lib/configuration.js'
```
It might be nicer to move this function into `@mdx-js/language-service`. Although it’s not ideal `loadPlugins` depends on Node.js APIs.

remcohaszing commented on packages/vscode-mdx/src/typescript-plugin.js:
```diff
@@ -0,0 +1,107 @@
+/**
```
I think it would be nice to publish this as `@mdx-js/typescript-plugin`, but we can do that later.

remcohaszing commented on packages/vscode-mdx/src/typescript-plugin.js:
```diff
@@ -0,0 +1,107 @@
+      return info.languageService
+    },
+    getExternalFiles(project, updateLevel = 0) {
+      if (updateLevel >= 1 || !externalFiles.has(project)) {
+        const newFiles = searchExternalFiles(ts, project, ['.mdx'])
+        externalFiles.set(project, newFiles)
+        project.refreshDiagnostics()
+      }
+
+      return externalFiles.get(project) ?? []
```
Just a little rewrite to help TypeScript infer types better.

```suggestion
      let files = externalFiles.get(project)

      if (updateLevel >= 1 || !files) {
        files = searchExternalFiles(ts, project, ['.mdx'])
        externalFiles.set(project, files)
        project.refreshDiagnostics()
      }

      return files
```

johnsoncodehk committed:
Update packages/vscode-mdx/src/typescript-plugin.js

Co-authored-by: Remco Haszing <remcohaszing@gmail.com>
Signed-off-by: Johnson Chu <johnsoncodehk@gmail.com>

johnsoncodehk committed:
Update packages/vscode-mdx/src/typescript-plugin.js

Co-authored-by: Remco Haszing <remcohaszing@gmail.com>
Signed-off-by: Johnson Chu <johnsoncodehk@gmail.com>

johnsoncodehk committed:
Update packages/vscode-mdx/src/typescript-plugin.js

Co-authored-by: Remco Haszing <remcohaszing@gmail.com>
Signed-off-by: Johnson Chu <johnsoncodehk@gmail.com>

johnsoncodehk committed:
Update packages/vscode-mdx/src/typescript-plugin.js

Co-authored-by: Remco Haszing <remcohaszing@gmail.com>
Signed-off-by: Johnson Chu <johnsoncodehk@gmail.com>

johnsoncodehk committed:
Update packages/vscode-mdx/src/typescript-plugin.js

Co-authored-by: Remco Haszing <remcohaszing@gmail.com>
Signed-off-by: Johnson Chu <johnsoncodehk@gmail.com>

johnsoncodehk committed:
Update packages/vscode-mdx/src/typescript-plugin.js

Co-authored-by: Remco Haszing <remcohaszing@gmail.com>
Signed-off-by: Johnson Chu <johnsoncodehk@gmail.com>

johnsoncodehk committed:
Upgrade to Volar 2.0 alpha (#356)

Volar 2 offers various small enhancements.

Co-authored-by: Remco Haszing <remcohaszing@gmail.com>

remcohaszing committed:
Rework virtual files (#363)

Before Volar, the ESM, JSX, and markdown parts of virtual files had to
map character by character, with parts shadowed by whitespace
characters. This was still how virtual files were built. However,
because of how Volar mappings work, this is unnecessary.

This change overhauls how virtual files are generated. This leads to
more flexibility of virtual files, meaning we can omit or inject
content.

This is already leveraged in several ways:

- Markdown nodes in MDX are represented by TypeScript fragments. This
  provides compatibility with JSX elements that require children on a
  type level.
- Markdown text nodes in MDX are represented by TypeScript JSX
  expressions containing an empty string.
- JavaScript chunks are represented in the virtual markdown files as
  HTML comments.

remcohaszing committed:
Disable formatting (#364)

Formatting embedded content causes too much unexpected behaviour in MDX
files.

Closes #355

remcohaszing committed:
Enhance how the Props type is used (#366)

If the Props type isn’t defined, it now falls back to an empty shape
instead of any. Also all keys are marked as readonly. As a result, props
are no longer represented using the string `Props`, but an inline
representation of the shape of the props.

Given that the upcoming support for an MDX layout also requires a
patched version of `Props`, this new setup is more consistent.

Also the `props` description now contains a link to the props page on
the MDX website.

Andarist committed:
Fix the README title in the `language-server` package (#367)

Signed-off-by: Mateusz Burzyński <mateuszburzynski@gmail.com>

remcohaszing committed:
Support MDX layout (#365)

Closes #362

remcohaszing committed:
Deduplicate logic for generating virtual content

`addOffset` now handles generating content based on the original content
and offsets.

remcohaszing committed:
Include empty brackets in the generated JSX chunk

This can be used for example for bracket matching.

remcohaszing commented:
Can you retarget this PR to https://github.com/mdx-js/mdx-analyzer? I think that would be better for visibility and discussions. I’m not even sure of the target repo of a PR can be changed.

remcohaszing committed:
Fix a typo in the watch file extensions

remcohaszing committed:
Add jsx to watch file extensions

remcohaszing committed:
Fix GitHub releases of vscode-mdx (#369)

The solution is based on
https://github.com/statelyai/xstate-tools/blob/8a5df5db0e96e097b51e58280ae6405f2a0f9f94/scripts/tag-extension.js

remcohaszing committed:
Dedent virtual markdown chunks (#370)

MDX allows to use indentation. In regular markdown however, indented
code represents a code block. To work around this, all virtual markdown
chunks are now dedented. This may change the semantic meaning of certain
markdown, but it doesn’t affect editor features as far as I’m aware.

github-actions[bot] committed:
Version Packages (#359)

Co-authored-by: github-actions[bot] <github-actions[bot]@users.noreply.github.com>

johnsoncodehk committed:
Merge remote-tracking branch 'upstream/main' into ts-plugin

johnsoncodehk committed:
use createAsyncTSServerPlugin

johnsoncodehk committed:
Update fixtures.code-workspace

johnsoncodehk committed:
support typescript extension features

johnsoncodehk commented:
Here it's: https://github.com/mdx-js/mdx-analyzer/pull/371
